File: bindings-python/tests/test_render_script_integration.py
import os
from pathlib import Path

import pytest


def _normalize_svg(svg: str) -> str:
    """Trim each line and drop blanks, mirroring the Rust `check_svg_snapshot`
    normalization in crates/pluot/tests/test_utils/snapshot_utils.rs."""
    return "\n".join(line.strip() for line in svg.splitlines() if line.strip())


@pytest.mark.asyncio
async def test_render_script_integration_python():
    fixtures_dir = os.environ.get("PLUOT_RENDER_SCRIPT_FIXTURES_DIR")
    if not fixtures_dir:
        pytest.skip(
            "PLUOT_RENDER_SCRIPT_FIXTURES_DIR not set; run "
            "scripts/test_python_render_script_integration.sh instead of "
            "pytest directly to exercise this test"
        )
    fixtures_dir = Path(fixtures_dir)

    script = (fixtures_dir / "render_script.py").read_text()
    canonical_svg = (fixtures_dir / "canonical.svg").read_text()

    driver_src = script

    namespace = {}
    exec(compile(driver_src, "<render_script.py>", "exec"), namespace)
    img = await namespace["main"]()

    assert _normalize_svg(img) == _normalize_svg(canonical_svg)

File: bindings-r/tests/testthat/test-render-script-integration.R
normalize_svg <- function(svg) {
  # Trim each line and drop blanks, mirroring the Rust `check_svg_snapshot`
  # normalization in crates/pluot/tests/test_utils/snapshot_utils.rs.
  lines <- trimws(strsplit(svg, "\n", fixed = TRUE)[[1]])
  paste(lines[nzchar(lines)], collapse = "\n")
}

test_that("executing the generated render_script.R reproduces the canonical SVG", {
  fixtures_dir <- Sys.getenv("PLUOT_RENDER_SCRIPT_FIXTURES_DIR")
  skip_if(
    identical(fixtures_dir, ""),
    "PLUOT_RENDER_SCRIPT_FIXTURES_DIR not set; run scripts/test_r_render_script_integration.sh instead of R CMD check directly"
  )

  script_path <- file.path(fixtures_dir, "render_script.R")
  canonical_path <- file.path(fixtures_dir, "canonical.svg")

  script_env <- new.env(parent = globalenv())
  sys.source(script_path, envir = script_env)
  img <- script_env$img

  canonical_svg <- paste(readLines(canonical_path), collapse = "\n")

  expect_equal(normalize_svg(img), normalize_svg(canonical_svg))
})
